runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -111,6 +112,25 @@ public void translate(final RunnerApi.Pipeline pipeline, SparkTranslationContext
     QueryablePipeline p =
         QueryablePipeline.forTransforms(
             pipeline.getRootTransformIdsList(), pipeline.getComponents());
+    for (PipelineNode.PTransformNode transformNode : p.getTopologicallyOrderedTransforms()) {
+      // Pre-scan pipeline to count which pCollections are consumed as inputs more than once so
+      // their corresponding RDDs can later be cached.
+      for (String inputId : transformNode.getTransform().getInputsMap().values()) {
+        context.incrementConsumptionCountBy(inputId, 1);
+      }
+      // Executable stage consists of two parts: computation and extraction. This means the result
+      // of computation is an intermediate RDD, which we might also need to cache.
+      if (transformNode.getTransform().getSpec().getUrn().equals(ExecutableStage.URN)) {
+        context.incrementConsumptionCountBy(
+            getExecutableStageIntermediateId(transformNode),
+            transformNode.getTransform().getOutputsMap().size());
+      }
+      for (String outputId : transformNode.getTransform().getOutputsMap().values()) {
+        RunnerApi.Components components = pipeline.getComponents();
+        WindowedValueCoder outputCoder = getWindowedValueCoder(outputId, components);
```

Comment:
Nit: I would just inline this line. (Or, alternatively, is there any reason context does not have a handle to components itself?)

Reply:
Done. (Mostly my reasoning here is that the translator is responsible for all of the messy pipeline reading, leaving `SparkTranslationContext` responsible only for execution-level details).